### PR TITLE
Improve API Docs for Invite to Group, Iss#8087

### PR DESCRIPTION
**Pull request by JTorr**
https://github.com/HabitRPG/habitica/issues/8087
### Changes

Update documentation for Invite to Group to more accurately reflect current structure of parameters and error types.

screen shot 2016-10-02 at 7 25 18 pm
screen shot 2016-10-02 at 7 25 28 pm

---

UUID: 1bf2957d-600e-4952-9e7b-d796cf99763d

### CODE REVIEW

File: website/server/controllers/api-v3/groups.js [hunk 1]
```diff
@@ -648,38 +648,64 @@ async function _inviteByEmail (invite, group, inviter, req, res) {
  *
  * @apiParam {String} groupId The group _id ('party' for the user party and 'habitrpg' for tavern are accepted)
  *
- * @apiParam {Array} [emails] Body parameter - An array of emails addresses to invite
+ * @apiParam {Array} [emails] Body parameter - An array of objects, each
```
Comment (crookedneighbor) [inline website/server/controllers/api-v3/groups.js]: I think the preferred way of writing this is:

``` js
/*
 * // other stuff
 * @apiParam {Object[]} [emails] Body parameter - An array of objects, each representing one email address to invite
 * @apiParam {String} emails.email The email address of the user being invited
 * @apiParam {String} [emails.name] The name of the user being invited
*/
```

File: website/server/controllers/api-v3/groups.js [hunk 2]
```diff
@@ -648,38 +648,64 @@ async function _inviteByEmail (invite, group, inviter, req, res) {
...
+ * email address. Key 'name' is optional, value is invitee's name.
  * @apiParam {Array} [uuids] Body parameter - An array of uuids to invite
  *
  * @apiParamExample {json} Emails
- *   {
- *     "emails": ["user-1@example.com", "user-2@exmaple.com"]
- *   }
+ * {
+ *    "emails": [
+ *      {"email": "user-1@example.com"},
```
Comment (crookedneighbor) [inline website/server/controllers/api-v3/groups.js:685]: Good catch on this.

File: website/server/controllers/api-v3/groups.js [hunk 3]
```diff
@@ -648,38 +648,64 @@ async function _inviteByEmail (invite, group, inviter, req, res) {
...
  *
- * @apiSuccess {Array} data The invites
+ * @apiSuccess {Array} data The invitees
  * @apiSuccess {Object} data[0] If the invitation was a user id, you'll receive back an object. You'll recieve one Object for each succesful user id invite.
  * @apiSuccess {String} data[1] If the invitation was an email, you'll receive back the email. You'll recieve one String for each successful email invite.
  *
- * @apiSuccessExample {json} Successful Response
+ * @apiSuccessExample {json} Successful Response with Emails
+ * {
+ *   "success": true,
```
Comment (crookedneighbor) [inline website/server/controllers/api-v3/groups.js]: I think we should provide some boiler plate at the top describing the success and notifications properties. No need to provide them in every response example.

File: website/server/controllers/api-v3/groups.js [hunk 4]
```diff
@@ -648,38 +648,64 @@ async function _inviteByEmail (invite, group, inviter, req, res) {
...
  *   }
  *
- * @apiError GroupNotFound The group could not be found
- * @apiError InvalidInvitationParams An error relating to the data sent in `emails` and/or `uuids`
- * @apiError TooManyInvites A max of 100 invites (combined emails and user ids) can be sent out at a time
+ * @apiError BadRequest Request body has missing or invalid content.
+ * @apiError NotAuthorized Missing API user or API key, or user already invited
+ * @apiError NotFound The group could not be found
+ *
+ * @apiErrorExample
```
Comment (crookedneighbor) [inline website/server/controllers/api-v3/groups.js]: Let's not include `apiErrorExample`s unless there's some really tricky case.

**Review by crookedneighbor — CHANGES_REQUESTED** [review]

File: website/server/controllers/api-v3/groups.js (hunk 1 above)
Comment (JTorr) [reply website/server/controllers/api-v3/groups.js]: Ah, thank you. Will fix.

**Comment by JTorr** [comment]
@crookedneighbor Please review changes. Thanks!

File: website/server/controllers/api-v3/groups.js (hunk 3 above)
Comment (crookedneighbor) [reply website/server/controllers/api-v3/groups.js]: Looks like this still needs to be addressed.

Comment (JTorr) [reply website/server/controllers/api-v3/groups.js]: I see that I missed the Notification property. Just to clarify, do you mean the success and notification properties should be at the top of the file (representing all of the Groups API calls), or together with the other properties (ex. line 676)

Comment (crookedneighbor) [reply website/server/controllers/api-v3/groups.js]: I'm saying that every request, regardless of group, has these properties. We don't need to put them in every success example. We should adjust the template to include some introductory information about requests. Not something that needs to be done in this PR, but you should remove the success and notifications keys from this example.

**Comment by crookedneighbor** [comment]
Just pushed a commit that illustrates how errors should be structured. https://github.com/HabitRPG/habitica/commit/2dc09586784366fcb139c377d3993c52ea5ac9da

Any error that would be repeated should be defined at the top of the appropriate file. For instance, the `GroupNotFound` error is defined here: https://github.com/HabitRPG/habitica/commit/2dc09586784366fcb139c377d3993c52ea5ac9da#diff-b950a7c557b35ea58bc2685532c46170R25

```
/**
 * @apiDefine GroupNotFound
 * @apiError (404) {NotFound} GroupNotFound The specified group could not be found.
 */
```

And it's used like this:

```
* @apiUse GroupNotFound
```

If it's an error that only occurs in a single route, it should be defined in that route:

```
 * @apiError (statusCode) {ErrorType} NameForError A description of the error
```

**Comment by JTorr** [comment]
@crookedneighbor Thanks for the example. Very helpful. LMK if these changes are what you were looking for.

**Comment by crookedneighbor** [comment]
What I'm saying regarding the `success` and `notifications` params returned in a successful response is that we _should not_ include them on every request. Instead, we should add information in the header of the API doc (see http://apidocjs.com/#headerfooter) about them. We should also include information about authenticating.

Those seem like separate concerns from this PR. So for now, just remove the information about `success` and `notifications`.

For the errors, I don't think we should generalize them. IE, let's be specific to the resource:

``` js
// only applicable to the invite route, so we just add it there
* @apiError (400) {BadRequest} NoEmailProvided An email address was not provided in the `emails` body param
// for group update and creation, so we define it at the top and use `@apiUse GroupBodyInvalid` in the routes
* @apDefine GroupBodyInvalid
* @apiError (400) {BadRequest} GroupBodyInvalid A parameter in the group body was invalid
```

Also, just opened a PR to correct our use of `NotAuthorized` which should only be used for authentication. Here's a link to the changes to the `invite` route: https://github.com/HabitRPG/habitica/pull/8103/files#diff-b950a7c557b35ea58bc2685532c46170R550

**Comment by JTorr** [comment]
@crookedneighbor How's it look now?

File: website/server/controllers/api-v3/groups.js [hunk 5]
```diff
@@ -669,38 +674,68 @@ async function _inviteByEmail (invite, group, inviter, req, res) {
...
- *     "emails": ["user-1@example.com", "user-2@exmaple.com"],
- *     "uuids": ["user-id-of-existing-user", "user-id-of-another-existing-user"],
- *   }
+ * {
+ *   "emails": [
+ *       {"email": "user-1@example.com"},
+ *       {"email": "user-2@example.com"}
+ *   ],
+ *   "uuids": ["user-id-of-existing-user"]
+ *}
```
Comment (crookedneighbor) [inline website/server/controllers/api-v3/groups.js]: Space here between `*` and `}`

File: website/server/controllers/api-v3/groups.js [hunk 6]
```diff
@@ -669,38 +674,68 @@ async function _inviteByEmail (invite, group, inviter, req, res) {
...
+ * @apiError (400) {BadRequest} UuidOrEmailOnly The `emails` and `uuids` params
+ * were both missing and/or a key other than `emails` or `uuids` was provided in
+ * the body param.
+ * @apiError (400) {BadRequest} CannotInviteSelf User id or email of invitee matches that
+ * of inviter.
+ * @apiError (400) {BadRequest} UserAlreadyInvited The user has already been
+ * invited to the group.
+ * @apiError (400) {BadRequest} UserAlreadyInGroup The user is already a member
+ * of the group.
+ * @apiError (400) {BadRequest} MustBeArray The `uuids` or emails body param is
```
Comment (crookedneighbor) [inline website/server/controllers/api-v3/groups.js]: emails should be surrounded in back ticks like uuids is.

File: website/server/controllers/api-v3/groups.js [hunk 7]
```diff
@@ -669,38 +674,68 @@ async function _inviteByEmail (invite, group, inviter, req, res) {
...
+ * not an array.
+ * @apiError (400) {BadRequest} UserNotFound One or more of the `uuids` does not
```
Comment (crookedneighbor) [inline website/server/controllers/api-v3/groups.js]: It's actually a `404 NotFound` error. You can just use the `UserNotFound` error defined in the user controller.

```
@apiUse UseNotFound
```

**Review by crookedneighbor — CHANGES_REQUESTED** [review]
We decided that those changes I made in the PR I linked would be a breaking change. So the errors need to stay as `NotAuthorized`.

Many of the `BadRequest` errors you have listed should actually be listed as `NotAuthorized`.

**Comment by JTorr** [comment]
@crookedneighbor Issues addressed.

**Review by crookedneighbor — APPROVED** [review]

**Comment by crookedneighbor** [comment]
Thanks for your patience. Notated toward your next tier.

**Pull request merged**